File: Week 1/problem-set-1/cash/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <string.h>

string checksum(long n);
string checkcard(string card, long credit);
int count_digits(long n);

// Luhn's algorithm
int main(void)
{
  
  long credit = get_long("Number: ");
  string card = checksum(credit);

  printf("%s\n", checkcard(card, credit));
}

string checksum(long n)
{
  int sum = 0;
  while (n > 0) 
  {
    int odd_digit = n % 10;
    n /= 10;
    int even_digit = n % 10;
    int two = 2*even_digit;
    if (two > 9)
    {
      sum += (two%10 + two/10);
    } else {
      sum += 2*even_digit + odd_digit;
    }
    n /= 10;
  }
  printf("%i\n", sum);
  if (sum%10 == 0){
    return "VALID";
  } else {
    return "INVALID";
  }
}

string checkcard(string card, long credit)
{  if (strcmp(card, "INVALID") == 0)
  {
    return "INVALID";
  }else {
    int length = count_digits(credit);

    if (length == 15){
      return "AMERICAN EXPRESS";
    } else if (length == 16){
      return "MASTERCARD";
    } else if ((length == 13 || length == 16)){
      return "VISA";
    }
  }
}

int count_digits(long n)
{
    int count = 0;
    while (n > 0)
    {
        n /= 10;
        count++;
    }
    return count;
}

File: Week 6/problem-set/sentimental-credit/credit.py
number = input("Number: ")
second_last = number[-2::-2]
another_last = number[-1::-2]
total = 0

for n in second_last:
    i = int(n) * 2
    if i > 9:
        total += i // 10 + i % 10
    else:
        total += i

for n in another_last:
    total += int(n)

valid = (total % 10 == 0)

if valid:
    numbers = len(number)
    digit = number[0:2]
    if digit in ['34', '37'] and numbers == 15:
        print("AMEX")
    elif digit in ['51', '52', '53', '54', '55'] and numbers == 16:
        print("MASTERCARD")
    elif number[0] == '4' and (numbers == 13 or numbers == 16):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

